luigi/contrib/pai.py

```diff
@@ -0,0 +1,300 @@
+                           headers={'Content-Type': 'application/json'}, data=request_json)
+        logger.debug('Get token response {0}'.format(response.text))
+        if response.status_code != 200:
+            msg = 'Get token request failed, response is {}'.format(response.text)
+            self.__logger.error(msg)
+            raise Exception(msg)
+        else:
+            self.__token = response.json()['token']
+
+    def __init__(self, *args, **kwargs):
+        """
+        :param pai_url: The rest server url of PAI clusters, default is 'http://127.0.0.1:9186'.
+        :param token: The toke used to auth the rest server of PAI.
+
+        """
+        super(PaiTask, self).__init__(*args, **kwargs)
+        self.__logger = logger
+        self.__init_token()
+
+    def __check_job_status(self):
+        response = rs.get(urljoin(self.__openpai.pai_url, '/api/v1/jobs/{0}'.format(self.name)))
+        self.__logger.debug('Check job response {0}'.format(response.text))
+        if response.status_code == 404:
+            self.__logger.debug('Job {0} is not found'.format(self.name))
+            return False
+        elif response.status_code != 200:
+            msg = 'Get job request failed, response is {}'.format(response.text)
+            self.__logger.error(msg)
+            raise Exception(msg)
+        job_state = response.json()['jobStatus']['state']
+        if job_state in ['UNKNOWN', 'WAITING', 'RUNNING']:
+            self.__logger.debug('Job {0} is running in state {1}'.format(self.name, job_state))
+            return False
+        else:
+            self.__logger.info('Job {0} finished in state {1}'.format(self.name, job_state))
+            return True
+
+    def run(self):
+        job = PaiJob(self.name, self.image, self.tasks)
+        job.virtualCluster = self.virtual_cluster
+        job.authFile = self.auth_file_path
+        job.codeDir = self.code_dir
+        job.dataDir = self.data_dir
+        job.outputDir = self.output_dir
+        job.retryCount = self.retry_count
+        job.gpuType = self.gpu_type
+        request_json = json.dumps(job,  default=slot_to_dict)
+        self.__logger.debug('Submit job request {0}'.format(request_json))
+        response = rs.post(urljoin(self.__openpai.pai_url, '/api/v1/jobs'),
+                           headers={'Content-Type': 'application/json',
+                                    'Authorization': 'Bearer {}'.format(self.__token)}, data=request_json)
+        self.__logger.debug('Submit job response {0}'.format(response.text))
+        # 202 is success for job submission, see https://github.com/Microsoft/pai/blob/master/docs/rest-server/API.md
+        if response.status_code != 202:
+            msg = 'Submit job failed, response code is {0}, body is {1}'.format(response.status_code, response.text)
+            self.__logger.error(msg)
+            raise Exception(msg)
+        while not self.__check_job_status():
```

Comment:
a 404 will keep this running indefinitely, is that expected?

Reply:
Not ready, I will split the method that checks the existence from self.__check_job_status()

Reply:
Still keep one method, but raise exeception for 404.